src/Components/Blazor/WebAssembly.Authentication/src/Services/RemoteAuthenticationService.cs

```diff
@@ -0,0 +1,188 @@
+        {
+            await EnsureAuthService();
+            var result = await _jsRuntime.InvokeAsync<RemoteAuthenticationResult<TRemoteAuthenticationState>>("AuthenticationService.signIn", context.State);
+            if (result.Status == RemoteAuthenticationStatus.Success)
+            {
+                UpdateUser(GetCurrentUser());
+            }
+
+            return result;
+        }
+
+        /// <inheritdoc />
+        public virtual async Task<RemoteAuthenticationResult<TRemoteAuthenticationState>> CompleteSignInAsync(
+            RemoteAuthenticationContext<TRemoteAuthenticationState> context)
+        {
+            await EnsureAuthService();
+            var result = await _jsRuntime.InvokeAsync<RemoteAuthenticationResult<TRemoteAuthenticationState>>("AuthenticationService.completeSignIn", context.Url);
+            if (result.Status == RemoteAuthenticationStatus.Success)
+            {
+                UpdateUser(GetCurrentUser());
+            }
+
+            return result;
+        }
+
+        /// <inheritdoc />
+        public virtual async Task<RemoteAuthenticationResult<TRemoteAuthenticationState>> SignOutAsync(
+            RemoteAuthenticationContext<TRemoteAuthenticationState> context)
+        {
+            await EnsureAuthService();
+            var result = await _jsRuntime.InvokeAsync<RemoteAuthenticationResult<TRemoteAuthenticationState>>("AuthenticationService.signOut", context.State);
+            if (result.Status == RemoteAuthenticationStatus.Success)
+            {
+                UpdateUser(GetCurrentUser());
+            }
+
+            return result;
+        }
+
+        /// <inheritdoc />
+        public virtual async Task<RemoteAuthenticationResult<TRemoteAuthenticationState>> CompleteSignOutAsync(
+            RemoteAuthenticationContext<TRemoteAuthenticationState> context)
+        {
+            await EnsureAuthService();
+            var result = await _jsRuntime.InvokeAsync<RemoteAuthenticationResult<TRemoteAuthenticationState>>("AuthenticationService.completeSignOut", context.Url);
+            if (result.Status == RemoteAuthenticationStatus.Success)
+            {
+                UpdateUser(GetCurrentUser());
+            }
+
+            return result;
+        }
+
+        /// <inheritdoc />
+        public virtual async ValueTask<AccessTokenResult> GetAccessToken()
+        {
+            await EnsureAuthService();
+            return await _jsRuntime.InvokeAsync<AccessTokenResult>("AuthenticationService.getAccessToken");
+        }
+
+        /// <inheritdoc />
+        public virtual async ValueTask<AccessTokenResult> GetAccessToken(AccessTokenRequestOptions options)
+        {
+            await EnsureAuthService();
+            return await _jsRuntime.InvokeAsync<AccessTokenResult>("AuthenticationService.getAccessToken", options);
+        }
+
+        /// <inheritdoc />
+        public virtual async Task<ClaimsPrincipal> GetCurrentUser()
```

Comment:
This looks a lot more expensive than I'd expect for a method that sounds like it's returning data we already have (i.e., it does a JS interop call and walks through JSON data to rebuild the principal from scratch).

Can we be caching the `ClaimsPrincipal` instead?

Reply:
updated to use internal caching